### ISSUE TITLE
Power9 installation

### ISSUE THREAD

**Issue by mikacashman** [rgcgithub/regenie#46]
Hi - I am **highly** interested in getting this tool installed on a power9 system.  Is there support for that?  

Bgen is installed just fine, but we have never been able to install plink (or plink2) on this system.  It looks like plink2 is embedded as a third-party tool in regenie which I think is the primary issue.  Can you help advise?  

The current error is suggesting we try a 32-bit version but I am not seeing how to flip this.  This is from source (begn path added to make -> make clean -> make).

```
g++ -O3 -Wall -ffast-math -std=c++11 -DVERSION_NUMBER=\"1.0.6.6\" -o src/Step1_Models.o -c src/Step1_Models.cpp -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0 -I./external_libs/pgenlib/ -I./external_libs/pgenlib//include/ -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/genfile/include/ -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/zstd-1.1.0/lib -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/db/include/ -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/sqlite3 -I./external_libs/  -fopenmp
In file included from ./external_libs/pgenlib/include/plink2_bits.h:23:0,
                 from ./external_libs/pgenlib/include/pgenlib_misc.h:77,
                 from ./external_libs/pgenlib/pgenlib_ffi_support.h:20,
                 from ./external_libs/pgenlib/pgenlibr.h:36,
                 from src/Geno.hpp:30,
                 from src/Step1_Models.cpp:29:
./external_libs/pgenlib/include/plink2_base.h:142:6: error: #error "64-bit builds currently require SSE2.  Try producing a 32-bit build instead."
 #    error "64-bit builds currently require SSE2.  Try producing a 32-bit build instead."
      ^
./external_libs/pgenlib/include/plink2_base.h:144:25: fatal error: emmintrin.h: No such file or directory
 #  include <emmintrin.h>
                         ^
compilation terminated.
make: *** [src/Step1_Models.o] Error 1

```

**Comment by chrchang** [comment]
I haven't tested on Power9, but this particular compilation restriction was removed from the plink2 codebase last month, so this may be worth retrying after updating the mirrored plink2 copy.

**Comment by joellembatchou** [comment]
@chrchang Thank you for clarifying this.

I have updated the PLINK2 library files (commit 0a1c8f52700ffb3a62f374f4b51c50dd6886b174), @mikacashman let me know if you have further issues.

Cheers,
Joelle

**Comment by mikacashman** [comment]
Thank you both so much for the quick feedback and updates it is highly appreciated!
On a fresh clone this is now the results of make

```
g++ -O3 -Wall -ffast-math -std=c++11 -DVERSION_NUMBER=\"1.0.6.7\" -o src/Step1_Models.o -c src/Step1_Models.cpp -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0 -I./external_libs/pgenlib/ -I./external_libs/pgenlib//include/ -I/gpfs/alpine/syb10     5/proj-shared/Personal/mcashman/Tools/summit/bgen -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/genfile/include/ -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/zstd-1.1.0/lib -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/db/include/ -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/sqlite3 -I./external_libs/  -fopenmp
In file included from ./external_libs/pgenlib/include/plink2_bits.h:23:0,
                 from ./external_libs/pgenlib/include/pgenlib_misc.h:77,
                 from ./external_libs/pgenlib/pgenlib_ffi_support.h:20,
                 from ./external_libs/pgenlib/pgenlibr.h:36,
                 from src/Geno.hpp:30,
                 from src/Step1_Models.cpp:29:
./external_libs/pgenlib/include/plink2_base.h:141:26: fatal error: x86/sse2.h: No such file or d     irectory
 #    include "x86/sse2.h"
                          ^
compilation terminated.
make: *** [src/Step1_Models.o] Error 1
```

**Comment by joellembatchou** [comment]
Sorry, I did not realize there were additional files needed from the PLINK2 repository. This has now been fixed (commit 00be179ad3b88b95d99fb5ff53a797867598108f)

Cheers,
Joelle

**Comment by mikacashman** [comment]
Thanks so much, now getting many BOOST complaints.

Obfuscated error is next.  Full error and the `/tmp/ccOawdeT.out` file attached at the end in case that was helpful.
```
g++ -O3 -Wall -ffast-math -std=c++11 -DVERSION_NUMBER=\"1.0.6.7\" -o src/Step1_Models.o -c src/Step1_Models.cpp -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0 -I./external_libs/pgenlib/ -I./external_libs/pgenlib//simde/ -I./external_libs/pgenlib//include/ -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/genfile/include/ -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/zstd-1.1.0/lib -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/db/include/ -I/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/sqlite3 -I./external_libs/  -fopenmp
In file included from /gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/mpl/aux_/integral_wrapper.hpp:22:0,
                 from /gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/mpl/int.hpp:20,
                 from /gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/mpl/lambda_fwd.hpp:23,
                 from /gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/mpl/aux_/na_spec.hpp:18,
                 from /gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/mpl/identity.hpp:17,
                 from /gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/iterator/detail/enable_if.hpp:11,
                 from /gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/iterator/transform_iterator.hpp:11,
                 from /gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/algorithm/string/iter_find.hpp:17,
                 from /gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/algorithm/string/split.hpp:16,
                 from src/Regenie.hpp:56,
                 from src/Step1_Models.cpp:28:
/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/concept_check.hpp: In function 'void boost::function_requires(Model*)':
/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/concept/detail/general.hpp:71:20: warning: typedef 'boost_concept_check45' locally defined but not used [-Wunused-local-typedefs]
       BOOST_PP_CAT(boost_concept_check,__LINE__)
                    ^
[...]
/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/concept/detail/general.hpp:71:7: note: in expansion of macro 'BOOST_PP_CAT'
       BOOST_PP_CAT(boost_concept_check,__LINE__)
       ^
/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/concept/assert.hpp:44:5: note: in expansion of macro 'BOOST_CONCEPT_ASSERT_FN'
     BOOST_CONCEPT_ASSERT_FN(void(*)ModelInParens)
     ^
/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/3rd_party/boost_1_55_0/boost/algorithm/string/find_format.hpp:262:13: note: in expansion of macro 'BOOST_CONCEPT_ASSERT'
             BOOST_CONCEPT_ASSERT((
             ^
In file included from ./external_libs/pgenlib//simde/x86/../simde-common.h:30:0,
                 from ./external_libs/pgenlib//simde/x86/mmx.h:30,
                 from ./external_libs/pgenlib//simde/x86/sse.h:33,
                 from ./external_libs/pgenlib//simde/x86/sse2.h:35,
                 from ./external_libs/pgenlib/include/plink2_base.h:141,
                 from ./external_libs/pgenlib/include/plink2_bits.h:23,
                 from ./external_libs/pgenlib/include/pgenlib_misc.h:77,
                 from ./external_libs/pgenlib/pgenlib_ffi_support.h:20,
                 from ./external_libs/pgenlib/pgenlibr.h:36,
                 from src/Geno.hpp:30,
                 from src/Step1_Models.cpp:29:
./external_libs/pgenlib//simde/x86/sse2.h: In function 'int32_t simde_mm_extract_epi16(simde__m128i, int)':
./external_libs/pgenlib//simde/x86/../hedley.h:832:60: sorry, unimplemented: unexpected AST of kind compound_literal_expr
 #  define HEDLEY_STATIC_CAST(T, expr) (static_cast<T>(expr))
                                                            ^
./external_libs/pgenlib//simde/x86/sse2.h:3063:9: note: in expansion of macro 'HEDLEY_STATIC_CAST'
     r = HEDLEY_STATIC_CAST(uint16_t, vec_extract(a_.altivec_i16, imm8));
         ^
./external_libs/pgenlib//simde/x86/sse2.h:3063: confused by earlier errors, bailing out
Preprocessed source stored into /tmp/ccOawdeT.out file, please attach this to your bugreport.
make: *** [src/Step1_Models.o] Error 1
```

[make_error.txt](https://github.com/rgcgithub/regenie/files/5461344/make_error.txt)
[ccOawdeT.out.txt](https://github.com/rgcgithub/regenie/files/5461345/ccOawdeT.out.txt)

**Comment by joellembatchou** [comment]
Which version of GCC are you using?

**Comment by mikacashman** [comment]
Ah always a good question.  I was using `gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-37)` but changing to `gcc (GCC) 9.1.0` managed to get it installed!

However, when I ran `test/test_bash.sh` I received the following.  I have yet to test on my own data yet.

```
Output
------
terminate called after throwing an instance of 'std::length_error'
  what():  basic_string::_S_create
phenotype 1 (Y1) : test/test_bash.sh: line 63: 15528 Aborted                 (core dumped) ./$reg      enie_bin $rgcmd
Step 1 of REGENIE did not finish successfully. Update to most recent REGENIE version (using 'git       pull') and re-compile the software (using 'make clean && make').
[mcashman@login2.summit regenie]$ git clone
You must specify a repository to clone.
```

**Comment by mikacashman** [comment]
**Update:** 

I also tried running the example from [here](https://rgcgithub.github.io/regenie/options/) and got a similar error:
```
Output
------
terminate called after throwing an instance of 'std::length_error'
  what():  basic_string::_S_create
phenotype 1 (Y1) : run_example.sh: line 20: 36252 Aborted                 (core dumped) $regenie --step 1 --bed $basepath/example/example --exclude $basepath/example/snplist_rm.txt --covarFile $basepath/example/covariates.txt --phenoFile $basepath/example/phenotype_bin.txt --remove $basepath/example/fid_iid_to_remove.txt --bsize 100 --bt --lowmem --lowmem-prefix output/tmp_rg --out output/fit_bin_out
run_example.sh: line 24: ./regenie: No such file or directory
```

And on some test data of our own a different message but still a core dump:
```
Fitting null model
 * bgen             : [/gpfs/alpine/syb105/proj-shared/Personal/mcashman/Data/GWAS_poplar/bgen1.1/869_NoCal_TagSNPs_CR0.99_imputed.bgen]
   -summary : bgen file (v1.1 layout, uncompressed) with 869 anonymous samples and 786069 variants.
regenie: /gpfs/alpine/syb105/proj-shared/Personal/mcashman/Tools/summit/bgen/genfile/include/genfile/bgen/bgen.hpp:504: bool genfile::bgen::read_snp_identifying_data(std::istream&, const genfile::bgen::Context&, std::string*, std::string*, std::string*, genfile::bgen::uint32_t*, NumberOfAllelesSetter, AlleleSetter) [with NumberOfAllelesSetter = BgenParser::read_variant(std::string*, uint32_t*, std::string*, std::vector<std::__cxx11::basic_string<char> >*)::<lambda(std::size_t)>; AlleleSetter = BgenParser::read_variant(std::string*, uint32_t*, std::string*, std::vector<std::__cxx11::basic_string<char> >*)::<lambda(std::size_t, const string&)>; std::istream = std::basic_istream<char>; std::string = std::__cxx11::basic_string<char>; genfile::bgen::uint32_t = unsigned int]: Assertion `0' failed.
run_regenie.sh: line 26: 62393 Aborted                 (core dumped) $regenie --step 1 --bsize 100 --bt --lowmem --covarFile $cpath --bgen $bgenpath --phenoFile $ppath --out ${opath}_step1
```

**Comment by joellembatchou** [comment]
Hi,

It will take me a bit of time to track the source of these errors on Power9 systems as I cannot replicate it on the Ubuntu/Max OSX systems I have access to (and from the logs there seems to be multiple sources of these errors). Is it possible for you to use the [Regenie docker image](https://github.com/rgcgithub?tab=packages&repo_name=regenie) provided in the repo? (or making your own using the provided [Dockerfile](https://github.com/rgcgithub/regenie/blob/master/Dockerfile)) ?

**Comment by mikacashman** [comment]
Docker isn't allowed on this power9 systems due to privileges.  But I can try to make a singularity image from the docker image and use that.  There is some concern there might still be complaints about ppc64le vs x86 differences depending on how the container is built, but I can certainly try it.

Are there any more logs/specs/or tests that would be helpful?  We do have some urgency for getting this tool running on this system.  I am happy to be helpful in any way I can.  Again I really appreciate all the swift and involved help I have gotten so far.

**Comment by mikacashman** [comment]
@joellembatchou another update.  I looked into running a singularity image of the docker build on our power9 system and it does not look possible.  There is no power9 system available for us to build the docker image to be able to then port it to singularity, and building it on x86 and then using on power9 won't work.

So unfortunately I don't see that being an option.  Do you have any further ideas?  I am happy to provide more logs/tests to help figure out the latest run errors.

**Comment by mikacashman** [comment]
I would like to update that I managed to get regenie operational on our power9 system.  I am unsure what the exact key fix was, but I have a suspicion it was an issue with boost during compilation.  I walked a colleague through the process and they were successful.  After updating a few modules to match theirs (including boost) I was also successful, perhaps an update with bgen as well as I rebuilt that.  So happy to report working!

Again I appreciate the help on this, we are thrilled to be able to use this toolkit now.
